### ISSUE TITLE
React ... `@ts-expect-error` in a multi-line JSX leads to false-positive type-checks

### ISSUE THREAD

**Issue by nico-w** [microsoft/TypeScript#49972]
# Bug Report

Using `@ts-expect-error` in a multi-line JSX leads to false-positive type-checks.

### 🔎 Search Terms

- jsx, ignore, expect, error

### 🕗 Version & Regression Information

`"typescript": "4.5.4"` & `"typescript": "4.7.4"`
I also tried `typescript@next` (`4.8.0-dev.20220720`)

This is the behavior in every version I tried, and I reviewed the FAQ for entries about JSX.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220720#code/JYWwDg9gTgLgBAbzgJQKYEMDGMA0cBiArgHbbATEDCE4Fqx8AvnAGZQ1wDkUG2nAUP1QAPSLDiYKAZ3gBZAJ7VaxejABcBEmQpLIKhgB4E-OHABu6ADbAAJgAV2YDTKjBiAc36MAfHAC8cAAUYI5SAJT+vsamPDCEUMRwBt4KunSGAPTeXoIiYvCSxDJwqTR6qgBMGkSkMORUZekwRibmVrYOEE5wLm6ePv5BIV3hkYitsfGJga2mybOmSaXKqguLbdb2jn6cFpudYALr6257HY4AjGuLp+1bXRXXWdcGy+UM16Zn92A73wdHY6mDIZOAAARgUgAtHlUNgYVB2FBPnBbvtLii0ecHnAQUkoVC4AALVA8VHwKREiCESw2Hp1SyWOAwInsADucHQxDplIgHJAqGZRPQ8BZwCkcGGYFREqxNie2SBr0UjX0MBR-22uzuAMxxE1XSuQOBoIh0Nh8NJSL1BrAjyBz3WBkdcDCXiAA)

### 💻 Code

```ts
import { React, FunctionComponent } from 'react'

export const MyComponent: FunctionComponent<{
  validProp: string
}> = (props) => {
  return <>MyComponent</>
}

export const MyComponent2: FunctionComponent<{
  validProp: string
}> = (props) => {
  return (
    <>
      <MyComponent
        validProp='validProp'
        invalidProp1
        invalidProp2
      />
      <MyComponent
        validProp='validProp'
        // @ts-expect-error
        invalidProp1
        invalidProp2 // <-- here it should still throw and show me that this prop is invalid
      />
      <MyComponent
        validProp='validProp'
        invalidProp1
        // @ts-expect-error
        invalidProp2
      />
    </>
  )
}
```

### 🙁 Actual behavior

Using `@ts-expect-error` if you want to suppress a warning for e.g. one invalid React component prop is not checking the rest of the other props anymore.

### 🙂 Expected behavior

Using `@ts-expect-error` if you want to suppress a warning for e.g. one invalid React component prop should still check the rest of the other props.

**Comment by RyanCavanaugh** [comment]
ts-expect-error runs *after* all typechecking has been completed and checks if an error was on that line. It has no effect on which errors get produced; it is not a suppression mechanism in the typechecker but rather a post-processing step you can use.

This is one of the caveats we tried to warn people about in #9448 🤷‍♂️😅

**Comment by nico-wellner** [comment]
Ok, very interesting. Thanks for your answer @RyanCavanaugh ! 😄 Would you mind helping me, finding the right way making sure that all lines of props are getting type-checked then? When I understood you correctly, I need to make sure that typescript is not "stopping" doing a following type-check on all other props and then I could use this suppression in "my correct manner"? Is that correct? 🤔

**Comment by nico-w** [comment]
Ok, very interesting. Thanks for your answer @RyanCavanaugh ! 😄 Would you mind helping me, finding the right way making sure that all lines of props are getting type-checked then? When I understood you correctly, I need to make sure that typescript is not "stopping" doing a following type-check on all other props and then I could use this suppression in "my correct manner"? Is that correct? 🤔

**Comment by RyanCavanaugh** [comment]
I don't understand what you're trying to do - are you writing a .d.ts test file?

**Comment by nico-w** [comment]
No. I'm trying to make my components props as type-safe as possible - with using `@ts-expect-error` for certain situations in which I want to suppress an expected error for _just one prop_. But all other props should still get properly type-checked.

Using `@ts-expect-error` in those situations is risky because it doesn't show you any TS error anymore that would potentially be in that component _aside_ of the one error you wanted to suppress.

In this example, I defined a component `MyComponent` and it has one valid prop named `validProp`. That means, whenever this component gets other props passed, it should fail with a TS error. So, in this example it should fail because of two extra props `invalidProp1` and `invalidProp2`. When I'm now suppressing the first invalid prop `invalidProp1` with `@ts-expect-error`, it should - in my opinion - still show me an error for the second invalid prop `invalidProp2` ... but it doesn't.

```tsx
const MyComponent: FunctionComponent<{
  validProp: string
}> = (props) => {
  return <>MyComponent</>
}

[...]

<MyComponent
  validProp='validProp'
  // @ts-expect-error
  invalidProp1
  invalidProp2 // <-- here it should still throw and show me that this prop is invalid
/>

[...]
```

And from your answer, I may understood it correctly, that I need to make sure that TS is checking all props and not just one at a time. In this screenshot from my IDE you can see that it's only checking the first invalid prop, but stops and doesn't check the second invalid prop. How can I change this? ... and if this is working, then I could potentially use `@ts-expect-error` to suppress just certain props. How do you think?

![image](https://user-images.githubusercontent.com/22443552/180143121-77241bf0-5c19-4a4a-b545-b6f54c2c6ed1.png)

Thank you.

Nico

**Comment by vsiao** [comment]
This looks to be a specific case of #47551
